### PR TITLE
Closes #8268: Show admin notice when Used-CSS subfolders lack write permissions

### LINKED ISSUE #8268: Show admin notice or log message when Used-CSS subfolders do not have writing permissions
**Describe the Issue**
We usually show an admin notice when WP Rocket folders, such as the used-css folder is not writable, but with the sub-folders that do not have write permission, no notice is shown.

An example case is a Used CSS subfolder getting created, but due to write permission, the Used CSS cannot be applied to the pages.

**Expected behavior**
Reference from @engahmeds3ed 

> We may need to enhance this to cover this case, or at least show log message related to it because it's confusing to save the message coming from the API response in error message when it succeeds.

**Additional context**
- Link to [Slack discussion](https://group-onecom.slack.com/archives/C08EFCZ68NP/p1775479138152379)

### PR DESCRIPTION
**Pull request by Miraeld**
> [!NOTE]
> This pull request was generated by the AI delivery pipeline (Claude Sonnet 4.6). All code has been reviewed by automated quality gates (DOD L1, unit tests, Lead Review, QA) before merge.

# Description

When a Used-CSS subfolder already exists but is not writable, WP Rocket previously stored an incorrect API-success message as the DB error reason and showed no admin notice. This PR surfaces the failure correctly at every layer — filesystem, database, and admin UI. Fixes #8268.

## What was done

Three coordinated fixes were applied across the RUCSS stack:

1. **`Jobs/Manager::process()`** — corrected the `make_status_failed()` call to pass the locally constructed `$message` instead of `$job_details['message']` (the API success payload). This is the root-cause one-line bug fix for the misleading DB error.
2. **`Controller/Filesystem::write_used_css()`** — added an explicit writability check on the target directory after `rocket_mkdir_p()` succeeds. When the directory exists but is not writable, the method now returns `false` immediately instead of attempting a doomed `put_contents()` call.
3. **`Controller/UsedCSS::notice_write_permissions()`** — extended to check a transient (`rocket_rucss_subfolder_permission_error_<blog_id>`) that `Manager::process()` sets on write failure. When the transient is present and RUCSS is enabled, an error notice is shown reusing the existing `rocket_notice_writing_permissions()` helper. The transient expires after 24 hours so the notice clears automatically once permissions are fixed.

Unit tests were added for the `process()` success/failure branches, the unwritable-directory scenario in `writeUsedCss`, and all four `noticeWritePermissions` scenarios (no error, top-level error, subfolder transient, RUCSS disabled).

## Type of change

- [ ] New feature (non-breaking change which adds functionality).
- [x] Bug fix (non-breaking change which fixes an issue).
- [ ] Enhancement (non-breaking change which improves an existing functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as before).
- [ ] Sub-task of #8268
- [ ] Chore
- [ ] Release

### What was tested

**Automated tests (code analysis + unit tests + WP-CLI smoke tests):**

- 63 RUCSS-group unit tests, 199 assertions — all pass (0 failures; 1 pre-existing risky flag from PHP 8.4 vfsStream deprecation, unrelated to this PR)
- All 3 new test files pass cleanly

**Per-criterion results:**

1. **Admin notice shown when subfolder lacks write permissions** — PASS. `UsedCSS::notice_write_permissions()` reads the blog-scoped transient and calls `rocket_notice_html(['status' => 'error', ...])` when set. Test `shouldShowSubfolderNoticeWhenTransientSet` confirms this path.

2. **Correct local error message stored in DB** — PASS. `make_status_failed()` now receives `$message` ('RUCSS: Could not write used CSS to the filesystem: ' . $url). Test `shouldFailAndSetTransientWhenWriteFails` asserts the exact expected string.

3. **24-hour transient `rocket_rucss_subfolder_permission_error_<blog_id>` set** — PASS. `set_transient('rocket_rucss_subfolder_permission_error_' . get_current_blog_id(), $row_details->url, DAY_IN_SECONDS)`. Unit test + WP-CLI roundtrip in live environment confirm key format and expiry.

4. **Notice suppressed when RUCSS disabled or no issue** — PASS. Tests `shouldDoNothingWhenRucssDisabledEvenIfTransientSet`, `shouldDoNothingWhenFolderWritableAndNoTransient`, `shouldDoNothingWhenUserCannotManageOptions` all pass with `rocket_notice_html` never called.

5. **Log/error message recorded on write failure** — PASS. `logger::error($message)` called in write-failure branch. Test asserts `logger->error()` called exactly once with expected message.

**Smoke tests (WP-CLI, live environment at localhost:8888):**
- Settings page: HTTP 200
- `notice_write_permissions` hook registered on `admin_notices` at priority 10 — confirmed
- Transient lifecycle (`set_transient` → `get_transient` → `delete_transient`) for key `rocket_rucss_subfolder_permission_error_1` — confirmed correct key format and value storage

### How to test

**Environment:** WordPress multisite or single-site install with WP Rocket active and RUCSS enabled.

1. Identify the Used-CSS subfolder path (e.g. `wp-content/cache/used-css/1/a/b/`).
2. Create a subfolder manually and remove its write permissions: `mkdir -p .../used-css/1/a/b && chmod 444 .../used-css/1/a/b`.
3. Trigger RUCSS processing for a URL whose hash maps to that subfolder.
4. Verify that the WP Rocket log records an error message that includes the failing URL (not an API success message).
5. Verify that the row in the RUCSS queue table shows an error status with the correct local error message (not the API success payload).
6. Navigate to any wp-admin page and confirm an admin notice is shown referencing the Used-CSS subfolder permission issue.
7. Restore write permissions (`chmod 755 .../used-css/1/a/b`) and wait 24 hours (or manually delete the transient `rocket_rucss_subfolder_permission_error_<blog_id>`); confirm the notice disappears.
8. Disable RUCSS; confirm no notice appears even when the transient exists.

### Acceptance criteria

- [x] When a Used-CSS subfolder exists but lacks write permissions, an admin notice is shown.
- [x] When a Used-CSS subfolder write fails, the correct local error message is stored in DB (not the API success payload).
- [x] A 24-hour transient `rocket_rucss_subfolder_permission_error_<blog_id>` is set on subfolder permission failure (blog-ID-scoped for multisite).
- [x] Admin notice does not appear when RUCSS is disabled or when there is no permission issue.
- [x] A log/error message is recorded when write fails.

### Affected Features & Quality Assurance Scope

- **RUCSS (Remove Unused CSS)** — background job processing, filesystem write layer, and admin notices.
- **Multisite** — transient is scoped per blog ID to prevent cross-site pollution.
- **Admin notices** — the existing write-permissions notice hook is extended; no new hooks are registered.

### Documentation

`Filesystem::write_used_css()` now checks `$this->filesystem->is_writable( dirname( $file ) )` after `rocket_mkdir_p()` returns true. If the directory is unwritable it returns `false` without attempting `put_contents()`.

`Manager::process()` already had a branch that ran when `write_used_css()` returned false. That branch now also calls `set_transient( 'rocket_rucss_subfolder_permission_error_' . get_current_blog_id(), $row_details->url, DAY_IN_SECONDS )`. The `make_status_failed()` call in that branch is fixed to use `$message` (the locally built string) rather than `$job_details['message']` (the API response payload).

`UsedCSS::notice_write_permissions()` gains a second check after the top-level folder guard: it reads the blog-ID-scoped transient and, if present, calls `rocket_notice_html()` with an error-level notice using `rocket_notice_writing_permissions()` on the subfolder path. The early-return for `! $this->is_enabled()` is preserved, so the notice is suppressed when RUCSS is off.

### New dependencies

None.

### Risks

None identified. All changes are contained within the RUCSS stack. The new transient write occurs only when a write failure is already detected, adding negligible overhead. The admin notice path is guarded by both the capability check and the `is_enabled()` guard already present in the method.

### Follow-up tickets

- [#8350](https://github.com/wp-media/wp-rocket/issues/8350) — Extract `is_subfolder_writable()` helper into reusable method on `Filesystem` (NICE_TO_HAVE)

# Mandatory Checklist

## Code validation

- [x] I validated all the Acceptance Criteria. If possible, provide screenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style

- [x] I wrote a self-explanatory code about what it does.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unnecessary complexity.
- [x] Output messages (errors, notices, logs) are explicit enough for users to understand the issue and are actionnable.

## Unticked items justification

None.

# Additional Checks
- [x] In the case of complex code, I wrote comments to explain it.
- [x] When possible, I prepared ways to observe the implemented system (logs, data, etc.).
- [x] I added error handling logic when using functions that could throw errors (HTTP/API request, filesystem, etc.)

### CODE REVIEW

File: inc/Engine/Optimization/RUCSS/Controller/Filesystem.php [hunk 1]
```diff
@@ -84,6 +84,10 @@ public function write_used_css( string $hash, string $used_css ): bool {
 			return false;
 		}
 
+		if ( ! $this->filesystem->is_writable( dirname( $file ) ) ) {
```
Comment (Miraeld) [inline inc/Engine/Optimization/RUCSS/Controller/Filesystem.php:87]: [LOW] The spec's Step 2 describes a public `is_subfolder_writable(string $hash): bool` helper on `Filesystem` that would expose the subfolder writability check as reusable public API. The implementation skips that method and puts the check inline in `write_used_css()` instead.

**Fix:** This is a simplification, not a defect — the inline check satisfies all required behaviour. Consider adding the helper method in a follow-up if callers outside `write_used_css()` ever need subfolder writability checks.

**Comment by Miraeld** [comment]
> [!NOTE]
> Generated by the AI delivery pipeline (lead-reviewer · Claude Sonnet 4.6).

## Code Review — Issue #8268 / Branch: fix/8268-subfolder-write-permissions

### Spec Compliance

| Spec item | Status | Notes |
|-----------|--------|-------|
| Step 1: Fix wrong error message in `Manager::process()` — pass `$message` not `$job_details['message']` | ✅ Done | Line 165: `make_status_failed(..., $message)` is correct |
| Step 2: Add `is_writable()` check after `rocket_mkdir_p()` in `Filesystem::write_used_css()` | ✅ Done | Inline check added at line 87; `is_subfolder_writable()` helper was skipped (acceptable simplification, see LOW finding) |
| Step 3: Set blog-ID-scoped transient on write failure in `Manager::process()` | ✅ Done | `set_transient('rocket_rucss_subfolder_permission_error_' . get_current_blog_id(), ...)` with `DAY_IN_SECONDS` |
| Step 4: Check transient in `UsedCSS::notice_write_permissions()` — show subfolder notice | ✅ Done | `get_transient` checked after top-level folder guard; uses `false !== $failing_url` (correct WordPress transient pattern) |
| Step 5: Tests — `Tests/Unit/.../Jobs/process.php` (new) | ✅ Done | Scenarios A (write success) and B (write failure) both covered with correct assertions |
| Step 5: Tests — `writeUsedCss.php` extended | ✅ Done | New test `testShouldReturnFalseWhenDirectoryExistsButNotWritable` added; uses Mockery to simulate `is_writable` returning false |
| Step 5: Tests — `noticeWritePermissions.php` (new) | ✅ Done | Fixtures cover all four required scenarios plus a fifth (user cannot manage options) |
| Edge case: Top-level folder not writable — existing behaviour unchanged | ✅ Done | Early `return` after top-level notice preserves original path |
| Edge case: Subfolder not writable — correct message stored in DB and transient set | ✅ Done | `$message` (not `$job_details['message']`) is now passed to `make_status_failed()` |
| Edge case: Permissions fixed after 24 h — notice disappears | ✅ Done | Transient uses `DAY_IN_SECONDS` expiry |
| Edge case: Multisite blog-ID scoping on transient key | ✅ Done | Both `Manager::process()` and `notice_write_permissions()` call `get_current_blog_id()` and append it to the key |
| Edge case: RUCSS disabled when transient exists — no notice | ✅ Done | `is_enabled()` early-return runs before transient check; test fixture `shouldDoNothingWhenRucssDisabledEvenIfTransientSet` confirms |
| Edge case: `gzencode` fails — same error path | ✅ Done | `write_used_css()` returns `false` before directory check when `$css` is falsy; same `process()` failure branch runs |
| Out of scope: No new hooks added directly | ✅ Correct | No `add_action`/`add_filter` added |
| Out of scope: No new `apply_filters()` calls | ✅ Correct | All existing bare `apply_filters()` calls in changed files are pre-existing |
| Out of scope: No `get_option()` introduced | ✅ Correct | Settings accessed via injected `Options_Data` |

### Findings

| File | Location | Criticality | Finding | Fix |
|------|----------|-------------|---------|-----|
| `inc/Engine/Optimization/RUCSS/Controller/Filesystem.php` | `write_used_css()` line 87 | LOW | The spec described a public `is_subfolder_writable(string $hash): bool` helper method to expose the check as reusable API. The implementation inlines the check directly in `write_used_css()` instead. The inline form is functionally complete and simpler. | Non-blocking: add `is_subfolder_writable()` as a follow-up only if future callers outside `write_used_css()` need it |

### Test Coverage

PASS

- `Tests/Unit/.../Jobs/process.php`: covers write-success (no transient, `make_status_completed` called) and write-failure (correct `$message` to `make_status_failed`, transient key format, `DAY_IN_SECONDS` expiry, logger receives error).
- `Tests/Unit/.../Controller/Filesystem/writeUsedCss.php`: existing happy-path tests untouched; new `testShouldReturnFalseWhenDirectoryExistsButNotWritable` correctly mocks `WP_Filesystem_Direct::is_writable` returning `false`.
- `Tests/Unit/.../Controller/UsedCSS/noticeWritePermissions.php`: five fixture scenarios cover all four spec scenarios plus the `cannot_manage_options` guard. `rocket_get_constant` is handled by the base `TestCase::setUp()` → `StubTrait::stubRocketGetConstant()` chain; no explicit mock needed in the test class.

**Overall: PASS**

No blockers. One LOW finding (missing optional helper method) is a non-blocking follow-up.

**Follow-ups** (LOW — non-blocking):
- `inc/Engine/Optimization/RUCSS/Controller/Filesystem.php`: Consider extracting the `is_writable(dirname($file))` check into a public `is_subfolder_writable(string $hash): bool` method as the spec described, so future callers can query subfolder writability without going through `write_used_css()`.

**Comment by Miraeld** [comment]
> [!NOTE]
> Generated by the AI delivery pipeline (qa-engineer · Claude Sonnet 4.6).

## Test Report — Closes #8268: Show admin notice when Used-CSS subfolders lack write permissions

**Branch:** fix/8268-subfolder-write-permissions

### Environment Boot
| Step | Result |
|------|--------|
| `gh pr checkout` | ✅ success |
| `bin/dev-up.sh` | ✅ exit 0 |
| `http://localhost:8888` reachable | ✅ HTTP 200 |

### Strategy Selection
| Strategy | Used | Reason |
|----------|------|--------|
| A — API/functional | ✅ | Backend PHP changes — direct code audit, WP-CLI transient verification, and hook registration check |
| B — Browser/UI | ➖ | No frontend file changes (no JS, CSS, or Twig templates touched) |
| C — Analysis fallback | ✅ | Unit test suite run against all three new test files |

### Acceptance Criteria

| Acceptance Criterion | Validation Method | Result | Evidence |
|----------------------|-------------------|--------|----------|
| When a Used-CSS subfolder exists but lacks write permissions, an admin notice is shown | Code audit + unit test | ✅ PASS | `UsedCSS::notice_write_permissions()` reads transient `rocket_rucss_subfolder_permission_error_{blog_id}`; when truthy, calls `rocket_notice_html(['status' => 'error', ...])`. Test `shouldShowSubfolderNoticeWhenTransientSet` passes, asserting `rocket_notice_html` is called with `status = error`. |
| When a Used-CSS subfolder write fails, the correct local error message is stored in DB (not the API success payload) | Code audit + unit test | ✅ PASS | `Manager::process()` diff: `make_status_failed()` now receives `$message` (`'RUCSS: Could not write used CSS to the filesystem: ' . $row_details->url`) instead of the former `$job_details['message']` (API payload). Test `shouldFailAndSetTransientWhenWriteFails` asserts the exact local error string is passed to `make_status_failed`. |
| A 24-hour transient `rocket_rucss_subfolder_permission_error_<blog_id>` is set on subfolder permission failure (blog-ID-scoped for multisite) | Code audit + unit test + WP-CLI | ✅ PASS | Code sets `set_transient('rocket_rucss_subfolder_permission_error_' . get_current_blog_id(), $row_details->url, DAY_IN_SECONDS)`. Test asserts key `rocket_rucss_subfolder_permission_error_1` and expiry `DAY_IN_SECONDS`. WP-CLI confirmed: key stored and retrieved correctly in the live environment as `rocket_rucss_subfolder_permission_error_1`. |
| Admin notice does not appear when RUCSS is disabled or when there is no permission issue | Code audit + unit test | ✅ PASS | Early return on `! $this->is_enabled()` preserved. Tests `shouldDoNothingWhenRucssDisabledEvenIfTransientSet`, `shouldDoNothingWhenFolderWritableAndNoTransient`, and `shouldDoNothingWhenUserCannotManageOptions` all pass — `rocket_notice_html` never called in those scenarios. |
| A log/error message is recorded when write fails | Code audit + unit test | ✅ PASS | `Manager::process()` calls `$this->logger::error($message)` before storing the error in DB. Test `shouldFailAndSetTransientWhenWriteFails` asserts `logger->error()` is called exactly once with the expected message. |

### Test Suite Results
All three new test files pass cleanly (9 tests, 37 assertions across all files):

| Test File | Tests | Result |
|-----------|-------|--------|
| `tests/Unit/inc/Engine/Optimization/RUCSS/Jobs/process.php` | 2 data sets | ✅ All pass |
| `tests/Unit/inc/Engine/Optimization/RUCSS/Controller/Filesystem/writeUsedCss.php` | 2 tests (happy path + unwritable dir) | ✅ All pass (1 risky: pre-existing vfsStream PHP 8.4 deprecation notice, not a test failure) |
| `tests/Unit/inc/Engine/Optimization/RUCSS/Controller/UsedCSS/noticeWritePermissions.php` | 5 data sets | ✅ All pass |

Full RUCSS group: 63 tests, 199 assertions — OK (1 pre-existing risky due to PHP 8.4 vfsStream deprecation, unrelated to this PR).

### Smoke Tests

| Area | Action | Result | Evidence |
|------|--------|--------|----------|
| Settings page | Navigated to `http://localhost:8888/wp-admin/options-general.php?page=wprocket` | ✅ PASS | HTTP 200, page loads without errors |
| notice_write_permissions hook registration | WP-CLI: inspected `$GLOBALS["wp_filter"]["admin_notices"]` | ✅ PASS | Hook `notice_write_permissions` registered at priority 10 |
| Transient lifecycle | WP-CLI set/get/delete `rocket_rucss_subfolder_permission_error_1` | ✅ PASS | Key matches spec, value stores URL, retrieval confirmed, cleanup successful |

### What Was Tested

**Code audit:** Reviewed full content of all three changed PHP files and the diff against base. Confirmed:
- `Filesystem::write_used_css()` guards with `is_writable()` after `rocket_mkdir_p()` succeeds, returning `false` immediately on unwritable directories.
- `Manager::process()` root-cause bug fix: `$message` (local error string) now passed to `make_status_failed()` instead of `$job_details['message']` (API success payload). Transient set only in the write-failure branch.
- `UsedCSS::notice_write_permissions()` logic was inverted in the original code (`if ( $this->filesystem->is_writable_folder() ) { return; }` → `if ( ! $this->filesystem->is_writable_folder() )` with the error notice inside the block). The new code correctly: (1) shows a top-level folder notice when `is_writable_folder()` returns false, (2) checks the subfolder transient after the top-level guard, (3) short-circuits when RUCSS is disabled.

**Test coverage assessment:** All five acceptance criterion branches are covered by unit tests. Both happy-path and failure-path are covered for `process()`. The unwritable-directory scenario in `write_used_css()` is explicitly tested with a mocked `WP_Filesystem_Direct`. All five notice scenarios (no error, top-level error, subfolder transient, RUCSS disabled, insufficient capability) are independently verified.

**Overall: PASS**

**Blockers:** None.

**Recommendations:**
- (NICE_TO_HAVE) The `writeUsedCss` fixture file includes a hardcoded partial CSS hash path (`95b0e3a1884eec34a989485f863ff.css.gz`) in the pre-built vfs structure that does not match the `shouldWriteUsedCss` hash (`fa2965d41f1515951de523cecb81f85e`). This causes the test to write to a new path rather than overwriting the fixture file — functionally correct but slightly misleading. Consider removing or aligning the stale pre-built directory in the fixture.

### Screenshots
Strategy B was not used — no frontend files changed (no JS, CSS, HTML, or Twig files in the diff).
